tests/ThreeWayMergeTest.php

```diff
@@ -0,0 +1,512 @@
+            I do not like doing it
+            it was easy'],
+        ];
+
+        $expected = [
+            'keyA' => ['This is fun
+            I do not like doing it
+            it was easy'],
+        ];
+        $merge = new ThreeWayMerge();
+        $result = $merge->performMerge($original, $local, $remote);
+        $this->assertEquals($expected, $result);
+    }
+    
+    public function testMultilineAdditionModification()
+    {
+        $original = [
+            'keyA' => ['This is not fun
+            I like doing it'],
+        ];
+
+        $local = [
+            'keyA' => ['This is fun
+            I do not like doing it'],
+        ];
+
+        $remote = [
+            'keyA' => ['This is not fun
+            I do not like doing it
+            it was easy'],
+        ];
+
+        $expected = [
+            'keyA' => ['This is fun
+            I do not like doing it
+            it was easy'],
+        ];
+        $merge = new ThreeWayMerge();
+        $result = $merge->performMerge($original, $local, $remote);
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testMultilineRemovalModification()
+    {
+        $original = [
+            'keyA' => ['This is not fun
+            I like doing it
+            it was not easy'],
+        ];
+
+        $local = [
+            'keyA' => ['This is fun
+            I do not like doing it
+            it was not easy'],
+        ];
+
+        $remote = [
+            'keyA' => ['This is not fun
+            I like doing it'],
+        ];
+
+        $expected = [
+            'keyA' => ['This is fun
+            I do not like doing it'],
+        ];
+        $merge = new ThreeWayMerge();
+        $result = $merge->performMerge($original, $local, $remote);
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testMultilineAdditionRemovalModificationTest1()
+    {
+        $original = [
+            'keyA' => ['This is not fun
+            I like doing it'],
+        ];
+
+        $local = [
+            'keyA' => ['This is fun
+            I like doing it
+            it was easy'],
+        ];
+
+        $remote = [
+            'keyA' => ['This is not fun'],
+        ];
+
+        $expected = [
+            'keyA' => ['This is fun
+            it was easy'],
+        ];
+        $merge = new ThreeWayMerge();
+        $result = $merge->performMerge($original, $local, $remote);
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testMultilineAdditionRemovalModificationTest2()
+    {
+        $original = [
+            'keyA' => ['This is not fun
+            I like doing it
+            it was easy'],
+        ];
+
+        $local = [
+            'keyA' => ['This is fun
+            I like doing it'],
+        ];
+
+        $remote = [
+            'keyA' => ['This is not fun'],
+        ];
+
+        $expected = [
+            'keyA' => ['This is fun
+            it was easy'],
+        ];
```

Comment:
We expect `it was easy` here because according to 3 way merge, we don't find what is in 2 nodes, we look for what has been modified because the person who modifies it knows what he is doing. and if 2 person modifies different lines on same nodes, we will merge the changes in final version on their respective lines. and if they've modified the same lines, then it'd return a conflict which needs to be solved manually. Here, `it was easy` was added by the developer on the local system and original and remote haven't made any change on that line. This is why we expect that line. You can go through http://www.drdobbs.com/tools/three-way-merging-a-look-under-the-hood/240164902 to clear the idea again. Please let me know if it's still not correct